src/auxiliary/mpi.jl

```diff
@@ -128,3 +128,10 @@ parallel execution of Trixi.jl.
 See the "Miscellaneous" section of the [documentation](https://docs.sciml.ai/DiffEqDocs/stable/basics/common_solver_opts/).
 """
 ode_unstable_check(dt, u, semi, t) = isnan(dt)
+
+# Custom MPI operators to work around
+# https://github.com/trixi-framework/Trixi.jl/issues/1922
+function reduce_vector_plus(x, y)
+    x .+ y
+end
+MPI.@Op(reduce_vector_plus, SVector)
```

Comment:
I don't think that will work...

You would need to say:
`MPI.@Op(reduce_vector_plus, SVector{3, Float32,})`

Reply:
See https://github.com/JuliaParallel/MPI.jl/blob/780aaa0fdb768713a329659338a9c9cde23c41a8/src/operators.jl#L84

Reply:
> I don't think that will work...
Absolutely right.

> Which means that we will have to do this for many types (different lengths, Float64 and maybe Float32, ...)
I was just trying to understand this. Is there no supertype?

Reply:
Sadly that doesn't work.

We generate a "wrapper" that looks like this:

```
function (w::OpWrapper{F,T})(_a::Ptr{Cvoid}, _b::Ptr{Cvoid}, _len::Ptr{Cint}, t::Ptr{MPI_Datatype}) where {F,T}
    len = unsafe_load(_len)
    @assert isconcretetype(T)
    a = Ptr{T}(_a)
    b = Ptr{T}(_b)
    for i = 1:len
        unsafe_store!(b, w.f(unsafe_load(a,i), unsafe_load(b,i)), i)
    end
    return nothing
end
```

So we get two pointer to an array of data, and we must reinterpret the pointer to a concrete type so that we can load it. Maybe one could use `t` to identify which Julia type one aught to use, but that would be less efficient.

Reply:
The type of code we can use here is basically like C, isn't it?

Reply:
If so, would it make sense to convert the `SVector` to play `Vector`s in our MPI routines to make our life easier and fix this issue?

Reply:
IIUC you currently have `data = Vector{SVector{5, Float64}}`, you could reinterpret that to `Ptr{Float64}` as long as your `reduce_vector` function is not using the fact that the datatype is a SVector.

Reply:
We do not need any special SVector functionality here. 
But we need to know the number of elements?

Reply:
Only when @vchuravy mentioned the `OpWrapper` I realized that it already iterates through something like a vector.
`_len` will be 1 in case of our `SVectors`, but carry the right number when using `Vectors` (where does this actually come from?). So, using a `Vector` or reinterpreting the `SVector` as `Ptr[Float64}` seems to make the reduction work, without a custom operator (currently tree 2d only).

Reply:
Doing this here now: #2067